README.md
# CIDR Matcher

Fast CIDR matcher. Given an input IPv4 or IPv6 address, it checks if it's inside a set of IP ranges, expressed in CIDR notation. This module is based upon the great [ip6addr](https://github.com/joyent/node-ip6addr) module.


## Installation

###  npm
```shell
npm install cidr-matcher
```

### git

```shell
git clone https://github.com/pracucci/node-cidr-matcher.git
```


## Usage

```js

var CIDRMatcher = require('cidr-matcher');

var matcher = new CIDRMatcher([ '2a05:d07c:2000:0:0:0:0:0/120', '192.168.1.0/24', '192.168.2.3/32', '192.168.3.2/32' ]);

matcher.contains('192.168.1.1'); // returns true
matcher.contains('192.168.1.2'); // returns true

matcher.contains('192.168.2.1'); // returns true
matcher.contains('192.168.2.2'); // returns false

matcher.contains('192.168.3.1'); // returns false
matcher.contains('192.168.3.2'); // returns false

matcher.containsAny([ '192.168.1.1', '192.168.1.2' ]); // return true
matcher.containsAny([ '192.168.2.1', '192.168.2.2' ]); // return true
matcher.containsAny([ '192.168.3.1', '192.168.3.2' ]); // return false

assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:0'));  // return true
assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:ff')); // return true
assert.ok(matcher.contains('2a05:d07c:3000:0:0:0:0:0'));  // return false

// You can also add / remove network classes on-the-fly
matcher.addNetworkClass('192.168.5.0/24');
matcher.contains('192.168.5.1'); // returns true

matcher.removeNetworkClass('192.168.5.0/24');
matcher.contains('192.168.5.1'); // returns false
```


## Benchmark

The following table shows the execution of `benchmark/` across different versions of this module and other modules you can find on npm. Each benchmark is **executed with a random set of 25000 IP addresses**, where each IP is checked against each network range (CIDR) in the test dataset.

| Module           | Version | Dataset               | Execution time |
| ---------------- | ------- | --------------------- | -------------- |
| _This one_       | `2.0.0` | AWS IPv4 (1385 CIDRs) | `228 ms`       |
| _This one_       | `2.0.0` | AWS IPv6 (474 CIDRs)  | `1799 ms`      |
| _This one_       | `1.0.5` | AWS IPv4 (1385 CIDRs) | `2895 ms` (_without IPv4 over IPv6 support_) |
| _This one_       | `1.0.5` | AWS IPv6 (474 CIDRs)  | _Unsupported_  |
| `is-in-subnet`   | `1.9.0` | AWS IPv4 (1385 CIDRs) | `96106 ms`     |
| `is-in-subnet`   | `1.9.0` | AWS IPv6 (474 CIDRs)  | `33482 ms`     |
| `ip-range-check` | `0.0.2` | AWS IPv4 (1385 CIDRs) | `390134 ms`    |
| `ip-range-check` | `0.0.2` | AWS IPv6 (474 CIDRs)  | `73083 ms`     |

_If you're aware of any other module that implements the IP in CIDR and you wanna see it benchmarked, please open an Issue or submit a PR._


## Contribute

### Run tests

```shell
npm test
```


### License

This software is licensed under the MIT License.

Copyright (c) 2015, Marco Pracucci <marco@pracucci.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGES.md
# Changes

### 2019-07-17 2.1.1
 - Upgraded `mocha`, addressing [code injection vulnerability](https://www.npmjs.com/advisories/813)

### 2019-02-20 2.1.0
 - Optimized IPv4 address against IPv4 CIDRs by a 20x factor

### 2019-02-20 2.0.0
 - **BREAKING CHANGE**: a network class must always be in CIDR notation (ending with `/number`) otherwise will throw an `Error`
 - **BREAKING CHANGE**: removed `Matcher.removeNetworkClass()`
 - Added IPv6 support to `Matcher`
 - Replaced `ip` module dependency with `ip6addr`

### 2017-03-30 1.0.5
 - IMPROVEMENT: replaced `chai` with `assert`, reducing devDependencies size (see [issue #4](https://github.com/pracucci/node-cidr-matcher/issues/4) - thanks to [christian-fei](https://github.com/christian-fei))

### 2017-03-30 1.0.4
 - FIX: correctly handle `0.0.0.0` IP address

### 2016-02-19 1.0.3
 - FIX: network and Broadcast addresses should match (issue #1)


test/MatcherTest.js
var assert  = require('assert'),
    Matcher = require('../src/Matcher.js');

describe('Matcher', function() {

    describe('contains()', function() {

        it('should return false with a bad input address', function() {
            var matcher = new Matcher([ '192.168.1.2/32', '2a05:d07c:2000::/40' ]);

            assert.ok(!matcher.contains('0.0.0.192.168.1.2'));
            assert.ok(!matcher.contains('xxx:d07c:2000::'));
        });

        it('should return true if an IPv4 matches a /32 network range', function() {
            var matcher = new Matcher([ '192.168.1.2/32' ]);

            assert.ok(!matcher.contains('192.168.1.1'));
            assert.ok(matcher.contains('192.168.1.2'));
            assert.ok(!matcher.contains('192.168.1.3'));
        });

        it('should return true if an IPv4 matches a /30 network range', function() {
            var matcher = new Matcher([ '192.168.1.0/30' ]);

            assert.ok(!matcher.contains('192.168.0.1'));

            assert.ok(matcher.contains('192.168.1.0'));
            assert.ok(matcher.contains('192.168.1.1'));
            assert.ok(matcher.contains('192.168.1.2'));
            assert.ok(matcher.contains('192.168.1.3'));
            assert.ok(!matcher.contains('192.168.1.4'));
        });

        it('should return true if an IPv4 matches a /24 network range', function() {
            var matcher = new Matcher([ '192.168.1.0/24' ]);

            assert.ok(!matcher.contains('192.168.0.1'));
            assert.ok(!matcher.contains('192.168.0.254'));

            assert.ok(!matcher.contains('192.168.2.1'));
            assert.ok(!matcher.contains('192.168.2.254'));

            assert.ok(matcher.contains('192.168.1.0'));
            assert.ok(matcher.contains('192.168.1.1'));
            assert.ok(matcher.contains('192.168.1.128'));
            assert.ok(matcher.contains('192.168.1.254'));
            assert.ok(matcher.contains('192.168.1.255'));
        });

        it('should return true if an IPv4 matches a /16 network range', function() {
            var matcher = new Matcher([ '192.168.1.0/16' ]);

            assert.ok(matcher.contains('192.168.0.0'));
            assert.ok(matcher.contains('192.168.0.1'));
            assert.ok(matcher.contains('192.168.0.254'));

            assert.ok(matcher.contains('192.168.128.1'));
            assert.ok(matcher.contains('192.168.128.254'));

            assert.ok(matcher.contains('192.168.255.1'));
            assert.ok(matcher.contains('192.168.255.254'));
            assert.ok(matcher.contains('192.168.255.255'));

            assert.ok(!matcher.contains('192.167.0.1'));
            assert.ok(!matcher.contains('192.169.0.1'));
        });

        it('should return true if an IPv4 matches multiple /32 network ranges', function() {
            var matcher = new Matcher([ '192.168.1.3/32', '192.168.1.2/32' ]);

            assert.ok(!matcher.contains('192.168.1.1'));
            assert.ok(matcher.contains('192.168.1.2'));
            assert.ok(matcher.contains('192.168.1.3'));
            assert.ok(!matcher.contains('192.168.1.4'));
        });

        it('should return true if an IPv4 matches network ranges', function() {
            var matcher = new Matcher([ '192.168.1.0/24', '192.168.2.3/32', '192.168.3.2/32' ]);

            assert.ok(matcher.contains('192.168.1.1'));
            assert.ok(matcher.contains('192.168.1.2'));
            assert.ok(matcher.contains('192.168.1.3'));

            assert.ok(!matcher.contains('192.168.2.1'));
            assert.ok(!matcher.contains('192.168.2.2'));
            assert.ok(matcher.contains('192.168.2.3'));

            assert.ok(!matcher.contains('192.168.3.1'));
            assert.ok(matcher.contains('192.168.3.2'));
            assert.ok(!matcher.contains('192.168.3.3'));
        });

        it('should return true if an IPv4 matches overlapping network ranges', function() {
            var matcher = new Matcher([ '192.168.1.0/24', '192.168.1.3/32', '192.168.1.2/32' ]);

            assert.ok(!matcher.contains('192.168.0.1'));
            assert.ok(!matcher.contains('192.168.0.254'));

            assert.ok(!matcher.contains('192.168.2.1'));
            assert.ok(!matcher.contains('192.168.2.254'));

            assert.ok(matcher.contains('192.168.1.1'));
            assert.ok(matcher.contains('192.168.1.2'));
            assert.ok(matcher.contains('192.168.1.3'));
            assert.ok(matcher.contains('192.168.1.128'));
            assert.ok(matcher.contains('192.168.1.254'));
        });

        it('should return true if an IPv4 matches a 0.0.0.0 range', function() {
            var matcher = new Matcher([ '0.0.0.0/24' ]);

            assert.ok(!matcher.contains('192.168.2.1'));
            assert.ok(!matcher.contains('192.168.2.254'));

            assert.ok(matcher.contains('0.0.0.0'));
            assert.ok(matcher.contains('0.0.0.10'));
            assert.ok(matcher.contains('0.0.0.254'));
        });

        it('should return true if an IPv6 matches a /128 network range', function() {
            var matcher = new Matcher([ '2a05:d07c:2000:0:0:0:0:1/128' ]);

            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:0:0'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:1'));
            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:0:2'));
        });

        it('should return true if an IPv6 matches a /120 network range', function() {
            var matcher = new Matcher([ '2a05:d07c:2000:0:0:0:0:1/120' ]);

            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:0'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:1'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:ff'));
            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:0:ffff'));
        });

        it('should return true if an IPv6 matches multiple /128 network ranges', function() {
            var matcher = new Matcher([ '2a05:d07c:2000:0:0:0:0:1/128', '2a05:d07c:2000:0:0:0:0:2/128' ]);

            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:0:0'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:1'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:2'));
            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:0:3'));
        });

        it('should return true if an IPv6 matches overlapping network ranges', function() {
            var matcher = new Matcher([ '2a05:d07c:2000:0:0:0:0:0/120', '2a05:d07c:2000:0:0:0:0:0/110' ]);

            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:4:0'));

            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:0'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:ff'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:1:0'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:3:ffff'));
        });

        it('should support mixed IPv4 and IPv6 network ranges', function() {
            var matcher = new Matcher([ '2a05:d07c:2000:0:0:0:0:0/120', '2a05:d07c:2000:0:0:0:0:0/110', '192.168.1.0/24', '192.168.2.3/32', '192.168.3.2/32' ]);

            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:4:0'));

            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:0'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:ff'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:1:0'));
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:3:ffff'));

            assert.ok(matcher.contains('192.168.1.1'));
            assert.ok(matcher.contains('192.168.1.2'));
            assert.ok(matcher.contains('192.168.1.3'));

            assert.ok(!matcher.contains('192.168.2.1'));
            assert.ok(!matcher.contains('192.168.2.2'));
            assert.ok(matcher.contains('192.168.2.3'));

            assert.ok(!matcher.contains('192.168.3.1'));
            assert.ok(matcher.contains('192.168.3.2'));
            assert.ok(!matcher.contains('192.168.3.3'));
        });

        it('should support IPv4 detection in a IPv6 network range', function() {
            var matcher = new Matcher([ '0:0:0:0:0:ffff:102:304/128' ]);

            assert.ok(!matcher.contains('1.2.3.3'));
            assert.ok(matcher.contains('1.2.3.4'));
            assert.ok(!matcher.contains('1.2.3.5'));
        });

        it('should support IPv6 detection in a IPv4 network range', function() {
            var matcher = new Matcher([ '1.2.3.4/32' ]);

            assert.ok(!matcher.contains('0:0:0:0:0:ffff:102:303'));
            assert.ok(matcher.contains('0:0:0:0:0:ffff:102:304'));
            assert.ok(!matcher.contains('0:0:0:0:0:ffff:102:305'));
        });

        it('should support IPv4 detection in a IPv6 network range spanning outside the IPv4 space', function() {
            var matcher = new Matcher([ '0:0:0:0:0:0:0:0/64' ]);

            assert.ok(matcher.contains('0.0.0.0'));
            assert.ok(matcher.contains('255.255.255.255'));

            assert.ok(matcher.contains('0:0:0:0:0:0:0:0'));
            assert.ok(matcher.contains('0:0:0:0:ffff:ffff:ffff:ffff'));
            assert.ok(!matcher.contains('0:0:0:1000:ffff:ffff:ffff:ffff'));
        });
    });


    describe('containsAny()', function() {

        it('should return true if any of input IPv4 addresses matches ranges', function() {
            var matcher = new Matcher([ '192.168.1.0/24', '192.168.2.3/32', '192.168.3.2/32' ]);

            assert.ok(!matcher.containsAny([ '192.168.2.1' ]));
            assert.ok(!matcher.containsAny([ '192.168.2.1', '192.168.3.3' ]));

            assert.ok(matcher.containsAny([ '192.168.2.1', '192.168.3.3', '192.168.2.3' ]));
        });

        it('should return true if any of input IPv6 addresses matches ranges', function() {
            var matcher = new Matcher([ '2a05:d07c:2000::/40' ]);

            assert.ok(!matcher.containsAny([ '2a05:d07c:1000:0:0:0:0:0' ]));
            assert.ok(!matcher.containsAny([ '2a05:d07c:1000:0:0:0:0:0', '2a05:d07c:3000:0:0:0:0:0' ]));

            assert.ok(matcher.containsAny([ '2a05:d07c:1000:0:0:0:0:0', '2a05:d07c:2000:0:0:0:0:0', '2a05:d07c:3000:0:0:0:0:0' ]));
        });
    });


    describe('addNetworkClass()', function() {

        it('should add an IPv4 cidr to the matcher', function() {
            var matcher = new Matcher();
            assert.ok(!matcher.contains('192.168.1.3'));

            matcher.addNetworkClass('192.168.1.1/24');
            assert.ok(matcher.contains('192.168.1.3'));
        });

        it('should add an IPv6 cidr to the matcher', function() {
            var matcher = new Matcher();
            assert.ok(!matcher.contains('2a05:d07c:2000:0:0:0:0:1'));

            matcher.addNetworkClass('2a05:d07c:2000::/40');
            assert.ok(matcher.contains('2a05:d07c:2000:0:0:0:0:1'));
        });

        it('should throw an error if the input network class is not in the CIDR notation', function() {
            var matcher = new Matcher();

            try {
                matcher.addNetworkClass('192.168.1.2');
                fail("should throw an Error");
            } catch(err) {
                assert.equal("Invalid argument: <addr>/<prefix> expected", err.message);
            }
        });
    });

});


src/Matcher.js
var ip6addr = require('ip6addr');


var Matcher = function(classes) {
    classes = classes || [];

    // Init
    this.ranges = {
        "ipv4": [],
        "ipv6": [],
    };

    // Import network classes
    for (var i = 0; i < classes.length; i++) {
        this.addNetworkClass(classes[i]);
    }
};

Matcher.prototype.addNetworkClass = function(input) {
    var cidr = ip6addr.createCIDR(input);

    // Detect if the input network class fits in the IPv4 address space.
    // If it doesn't an Error will be thrown and we'll add it to the IPv6
    // ranges in the catch block.
    try {
        var firstNumericAddr = cidr.address().toLong();
        var lastNumericAddr  = cidr.broadcast().toLong();

        this.ranges["ipv4"].push({
            cidr:  cidr,
            first: firstNumericAddr,
            last:  lastNumericAddr,
        });
    } catch(err) {
        // We've been unable to convert first/last address it to Long,
        // so it means it spans over the IPv6 address space
        this.ranges["ipv6"].push({
            cidr: cidr
        });
    }
};

Matcher.prototype.contains = function(addr) {
    // Parse input address
    try {
        addr = ip6addr.parse(addr);
    } catch(err) {
        return false;
    }

    var ipv4Ranges = this.ranges["ipv4"];
    var ipv6Ranges = this.ranges["ipv6"];

    // Check if the input address is within any IPv4 network range.
    // We detect if the input address is an IPv4. If yes, we can run an optimized
    // version which compares the IPv4 network ranges by numeric comparison.
    if (ipv4Ranges.length > 0) {
        if (addr.kind() === "ipv4") {
            var numericAddr = addr.toLong();

            for (var i = 0, length = ipv4Ranges.length; i < length; i++) {
                if (ipv4Ranges[i].first <= numericAddr && ipv4Ranges[i].last >= numericAddr) {
                    return true;
                }
            }
        } else {
            for (var i = 0, length = ipv4Ranges.length; i < length; i++) {
                if (ipv4Ranges[i].cidr.contains(addr)) {
                    return true;
                }
            }
        }
    }

    // Check if the input address is within any IPv6 network range
    if (ipv6Ranges.length > 0) {
        for (var i = 0, length = ipv6Ranges.length; i < length; i++) {
            if (ipv6Ranges[i].cidr.contains(addr)) {
                return true;
            }
        }
    }

    return false;
};

Matcher.prototype.containsAny = function(addrs) {
    for (var i in addrs) {
        if (addrs.hasOwnProperty(i) && this.contains(addrs[i])) {
            return true;
        }
    }

    return false;
};

module.exports = Matcher;


benchmark/benchmark.js
const CIDRMatcherMaster = require("../src/Matcher");
const CIDRMatcher105    = require("cidr-matcher");
const { isInSubnet }    = require("is-in-subnet");
const ipRangeCheck      = require("ip-range-check");


function generateRandomIPv4() {
    return "" +
        Math.floor(Math.random() * 255) + "." +
        Math.floor(Math.random() * 255) + "." +
        Math.floor(Math.random() * 255) + "." +
        Math.floor(Math.random() * 255);
}

function generateRandomIPv6() {
    const parts = [... new Array(8)].map(() => {
        return Number(Math.floor(Math.random() * 65535)).toString(16);
    });

    return parts.join(":");
}

function trackExecutionTimeMs(func) {
    const startTime = process.hrtime();

    func();

    const elapsedTime   = process.hrtime(startTime);
    const elapsedTimeMs = (elapsedTime[0] * 1000) + Math.round(elapsedTime[1] / 1000000);

    return elapsedTimeMs;
}


// Load Amazon CIDRs
const cidrsV4 = require("./amazon-cidrs.json").prefixes.map((item) => item.ip_prefix);
const cidrsV6 = require("./amazon-cidrs.json").ipv6_prefixes.map((item) => item.ipv6_prefix);

// Generate input fixtures
const numFixtures = 25000;
const randomV4Set = [... new Array(numFixtures)].map(() => generateRandomIPv4());
const randomV6Set = [... new Array(numFixtures)].map(() => generateRandomIPv6());


// Print datasets info
console.log("DATASETS");
console.log(`- ipv4: contains ${cidrsV4.length} CIDRs`);
console.log(`- ipv6: contains ${cidrsV6.length} CIDRs`);
console.log("");

// Run benchmark
function runBenchmark(moduleName, moduleVersion, datasetName, func) {
    const cidrs = datasetName === "ipv4" ? cidrsV4 : cidrsV6;
    const ips   = datasetName === "ipv4" ? randomV4Set : randomV6Set;

    const executionTime = trackExecutionTimeMs(() => {
        func(cidrs, ips);
    });

    console.log(`${moduleName} v. ${moduleVersion} against ${datasetName} dataset: ${executionTime} ms`);
}

console.log("BENCHMARK");

runBenchmark("node-cidr-matcher", "master", "ipv4", (cidrs, ips) => {
    let matcher = new CIDRMatcherMaster(cidrs);
    for (let i = 0; i < ips.length; i++) {
        matcher.contains(ips[i]);
    }
});

runBenchmark("node-cidr-matcher", "1.0.5", "ipv4", (cidrs, ips) => {
    let matcher = new CIDRMatcher105(cidrs);
    for (let i = 0; i < ips.length; i++) {
        matcher.contains(ips[i]);
    }
});

runBenchmark("node-cidr-matcher", "master", "ipv6", (cidrs, ips) => {
    let matcher = new CIDRMatcherMaster(cidrs);
    for (let i = 0; i < ips.length; i++) {
        matcher.contains(ips[i]);
    }
});

runBenchmark("is-in-subnet", "1.9.0", "ipv4", (cidrs, ips) => {
    for (let i = 0; i < ips.length; i++) {
        isInSubnet(ips[i], cidrs);
    }
});

runBenchmark("is-in-subnet", "1.9.0", "ipv6", (cidrs, ips) => {
    for (let i = 0; i < ips.length; i++) {
        isInSubnet(ips[i], cidrs);
    }
});

runBenchmark("ip-range-check", "0.0.2", "ipv4", (cidrs, ips) => {
    for (let i = 0; i < ips.length; i++) {
        ipRangeCheck(ips[i], cidrs);
    }
});

runBenchmark("ip-range-check", "0.0.2", "ipv6", (cidrs, ips) => {
    for (let i = 0; i < ips.length; i++) {
        ipRangeCheck(ips[i], cidrs);
    }
});
